### PR TITLE
KAFKA-9381: Fix publishing valid scaladoc for streams-scala

### PR DESCRIPTION
**Pull request by bbejeck**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -1439,7 +1440,7 @@ project(':streams:streams-scala') {
   }
 
   jar {
-    dependsOn 'copyDependantLibs'
+    dependsOn 'copyDependantLibs' , 'scaladocJar'
```
Comment (ijuma) [inline build.gradle]: Hmm, seems weird to require scaladoc when building jars.

Comment (bbejeck) [reply build.gradle]: @ijuma good point, I've removed it and still get the required doc jars

**Comment by bbejeck** [comment]
Results of rendering contents of scaladoc jar

Screen Shot 2020-10-23 at 2 59 32 PM

**Comment by ijuma** [comment]
We probably only want to publish scaladoc for kafka streams since it's the only module that has a public Scala API.

**Comment by mjsax** [comment]
@ijuma Correct! That is what KAFKA-9381 actually is all about.

LGTM.

**Comment by bbejeck** [comment]
>We probably only want to publish scaladoc for kafka streams since it's the only module that has a public Scala API.

ack, working on that now

File: build.gradle [hunk 2]
```diff
@@ -942,6 +942,7 @@ project(':core') {
     from(project(':streams').jar) { into("libs/") }
     from(project(':streams').configurations.runtime) { into("libs/") }
     from(project(':streams:streams-scala').jar) { into("libs/") }
+    from(project(':streams:streams-scala').configurations.archives.artifacts.files.filter { file -> file.name.matches(".*scaladoc.*")}) { into("libs/") }
```
Comment (mjsax) [inline build.gradle]: Do we actually need to do this? I don't see any other line here that would copy JavaDocs-jar into the `lib` directory?

\cc @ijuma

**Comment by bbejeck** [comment]
Need to test locally that the `uploadArchives` task triggers `archive` blocks in the build.gradle file.  If correct then the `kafka-streams-scalaXXXX-scaladoc.jar` should get created properly, and with scaladoc disabled for core, we won't publish non-public scaladoc

**Comment by bbejeck** [comment]
I was able to test this locally by setting the `mavenUrl` variable in `~/.gradle/gradle.properties` file to `/Users/bill/Downloads/testRepo/` on my laptop.  

Running the `uploadArchives` task locally produced these scaladoc jar files.

```
find . -name '*scaladoc.jar' | xargs ls -la
-rw-r--r--  1 bill  staff  1334323 Oct 26 10:48 ./org/apache/kafka/kafka-streams-scala_2.12/2.8.0-SNAPSHOT/kafka-streams-scala_2.12-2.8.0-20201026.144817-1-scaladoc.jar
-rw-r--r--  1 bill  staff  1326180 Oct 26 10:50 ./org/apache/kafka/kafka-streams-scala_2.13/2.8.0-SNAPSHOT/kafka-streams-scala_2.13-2.8.0-20201026.145048-1-scaladoc.jar
-rw-r--r--  1 bill  staff     9585 Oct 26 10:47 ./org/apache/kafka/kafka_2.12/2.8.0-SNAPSHOT/kafka_2.12-2.8.0-20201026.144723-1-scaladoc.jar
-rw-r--r--  1 bill  staff     9585 Oct 26 10:50 ./org/apache/kafka/kafka_2.13/2.8.0-SNAPSHOT/kafka_2.13-2.8.0-20201026.145017-1-scaladoc.jar
```
So the changes in this PR will produce a valid scaladoc jar file `streams-scala`, which has a public Scala API, but not for core, which is what we want.

So I think this PR is good now.

ping @ijuma, @mjsax  for another review

**Comment by ijuma** [comment]
What's in the following two files?

```text
-rw-r--r--  1 bill  staff     9585 Oct 26 10:47 ./org/apache/kafka/kafka_2.12/2.8.0-SNAPSHOT/kafka_2.12-2.8.0-20201026.144723-1-scaladoc.jar
-rw-r--r--  1 bill  staff     9585 Oct 26 10:50 ./org/apache/kafka/kafka_2.13/2.8.0-SNAPSHOT/kafka_2.13-2.8.0-20201026.145017-1-scaladoc.jar
```

Can we disable them altogether?

**Comment by bbejeck** [comment]
> What's in the following two files?

Just the manifest and the license 

```
jar tf ./org/apache/kafka/kafka_2.13/2.8.0-SNAPSHOT/kafka_2.13-2.8.0-20201026.145017-1-scaladoc.jar
META-INF/
META-INF/MANIFEST.MF
LICENSE
NOTICE
```
I'd have to look into what it would take to disable creating the scaladoc jar file, but since it's scala project I'm not sure.

**Comment by bbejeck** [comment]
@ijuma 

This latest change gets us where I think we want to be.  

Using the same test procedure from before, we only have scaladoc files for streams-scala.

```
find . -name '*scaladoc.jar' | xargs ls -la
-rw-r--r--  1 bill  staff  1334323 Oct 26 11:53 ./org/apache/kafka/kafka-streams-scala_2.12/2.8.0-SNAPSHOT/kafka-streams-scala_2.12-2.8.0-20201026.155354-1-scaladoc.jar
-rw-r--r--  1 bill  staff  1326180 Oct 26 11:56 ./org/apache/kafka/kafka-streams-scala_2.13/2.8.0-SNAPSHOT/kafka-streams-scala_2.13-2.8.0-20201026.155648-1-scaladoc.jar
```

The `build.gradle` file now only generates scaladoc archives specifically for streams-scala vs. all sub-projects using the scala plugin.  
In the future, if we need to generate public scaladoc for another module, we'll have to make the same change to that project as we did for streams-scala.

File: build.gradle [hunk 3]
```diff
@@ -731,6 +728,9 @@ project(':core') {
   println "Building project 'core' with Scala version ${versions.scala}"
 
   apply plugin: 'scala'
+  scaladoc {
+      enabled = false
+  }
```
Comment (ijuma) [inline build.gradle:738]: Do we still need this with your latest change? Also, can you please check the indenting generally? It should be two spaces.

Comment (bbejeck) [reply build.gradle:738]: >Do we still need this with your latest change?

I thought about taking it out, but I left it in as a fail-safe, thinking that if the build file changes somehow enabling scaladoc for all submodules, `core` will only provide the scaladoc jar file with the license and manifest.

I don't have a strong opinion on this, so I can remove if you want me to.

Ack on the indenting

**Comment by bbejeck** [comment]
updated this \cc @ijuma

File: build.gradle (hunk 3 above)
Comment (ijuma) [reply build.gradle:738]: ok, let's maybe add a comment explaining it then.

Comment (bbejeck) [reply build.gradle:738]: ack

**Comment by bbejeck** [comment]
updated this PR

\cc @ijuma

File: build.gradle (hunk 3 above)
Comment (ijuma) [reply build.gradle:738]: I tweaked it a bit

**Review by ijuma — APPROVED** [review]
LGTM

**Comment by bbejeck** [comment]
test failure unrelated and is a known flaky test `org.apache.kafka.streams.integration.EosBetaUpgradeIntegrationTest.shouldUpgradeFromEosAlphaToEosBeta[true]`

**Comment by bbejeck** [comment]
Merged #9486 into trunk

**Comment by bbejeck** [comment]
cherry-picked to 2.7

**Pull request merged**